feat(navbar): refactor navbar to support horizontal nav item placement and more

ocBruno opened the pull request:
fix #225

**Changes proposed in this pull request:**
- Refactor navbar function components 
- Abstract nav items to NavItem interface with a string type property that is the kebab-case version of the component NavIcon | NavHeader | NavLink | NavSearch | NavLinkList
- Support custom horizontal component position 
- Suggest new default theme and dark theme color
- Refactor tests to improve naming convention 
- Add className prop to each component which corresponds to its inner top level element's className to better support customization
- Add example with icons
- Design/layout changes in stories


I'm still willing to fix anything in this issue before approval if necessary or increment any changes. Also willing to update the frontend navbar to support these changes after approval!

ocBruno commented:
> Could you explain these changes a little bit more? I'm conflicted whether or not they are necessary. This definitely gives us a more flexible design, however, I'm not sure if that is something that we want. Since we know that our Navbar is always going to have some sort of brand, some sort of search then will have elements to the right of the Brand, and to the right of Search. If that is the case, then it seems like the original design ensures that the user of the component always gets the elements in the right place.

@jackcmeyer I don't think they are necessary but I think considering the original layouts can still be applied, the issue is resolved and now there are new positioning possibilities it gives the framework user more freedom.

ocBruno commented:
I'm not sure if this is a storybook bug or this is solvable with a feature but the children property in LinkList items aren't rendering the correct code preview. When I comment out the link list everything renders correctly so I'm not sure if anyone knows of a workaround, if I get some extra time I'll create a reproducible example and open a new issue in storybook repo considering I didn't find anything related.

I added a couple suggestions considering the colors. I suggest making the default theme light consdering the color white's association with hospitals and I believe it fits better for the default theme. 

In hospital environments, a colour preference for “whiteness” or pale
shades was found among patients with its connotations of cleanliness and
hygiene [2.14].
http://www.wales.nhs.uk/sites3/Documents/254/B(01)02%20Lighting%20and%20colour.pdf

I also suggested a change in the dark color, reducing brightness and saturation to cause less eye strain while maintaining readibility

jackcmeyer commented:
@ocBruno could we make the text of the nav links match the color of the brand and then be their current color when they are hovered over?

Plus, can the search bar remain to the left?

Can you provide a storybook example of content in the nav bar being justified to the right?

I'd like to see a navbar that is like

```
|[brand] [link1][link2]                [search][i1][i2]|
```

Where i1 and i2 are some icon such as a settings gear and profile icon.

ocBruno commented:
> @ocBruno could we make the text of the nav links match the color of the brand and then be their current color when they are hovered over?
> 
> Plus, can the search bar remain to the left?
> 
> Can you provide a storybook example of content in the nav bar being justified to the right?
> 
> I'd like to see a navbar that is like
> 
> ```
> |[brand] [link1][link2]                [search][i1][i2]|
> ```
> 
> Where i1 and i2 are some icon such as a settings gear and profile icon.

@jackcmeyer,
I added the example as provided and added support for classes in the nav items for better customization. 
As in leaving the search bar to the left do you mean in the examples 
I'm not sure about how we should go about matching the nav link color to the brand considering the various possibilities of color choice it could lead to some color or contrast conflicts. I think using the dark variation of the theme and letting the framework user modify the color themselves might be a good alternative.

> @ocBruno could we make the text of the nav links match the color of the brand and then be their current color when they are hovered over?
> 
> Plus, can the search bar remain to the left?
> 
> Can you provide a storybook example of content in the nav bar being justified to the right?
> 
> I'd like to see a navbar that is like
> 
> ```
> |[brand] [link1][link2]                [search][i1][i2]|
> ```
> 
> Where i1 and i2 are some icon such as a settings gear and profile icon.

Hey @jackcmeyer, thanks for the feedback once again.
I included the example and added the icons as requested and I also made className passable from component to it's parent element to improve customization.

As in maintaining the search bar to the left do you mean in the provided example or in one of the existing examples?

I'm not sure how we would go about adding the color according to the brand considering there could be some color or contrast conflicts. I think a good alternative is using the dark variant and letting the user customize any custom colors using css.

jackcmeyer commented:
> As in maintaining the search bar to the left do you mean in the provided example or in one of the existing examples?

@ocBruno Sorry, I meant the search bar to the right ☹️. I shouldn't look at Pull Requests late in the evening! The most recent example that you provided with icons looks perfect!

jackcmeyer commented:
@ocBruno Please let us know when this is no longer work in progress by taking the WIP out of the title! This is starting to look really good! I'm excited for this upgrade to reach master so we can implement these changes in the app.

ocBruno commented:
> @ocBruno Please let us know when this is no longer work in progress by taking the WIP out of the title! This is starting to look really good! I'm excited for this upgrade to reach master so we can implement these changes in the app.

Thanks for all the feedback and guidance @jackcmeyer!  I was waiting for some approval before finishing to make sure everyone agrees all on the changes but this issue and #233 should be finished by this weekend.

jackcmeyer commented on src/components/Navbar/Navbar.tsx:
```diff
@@ -5,86 +5,124 @@ import Form from 'react-bootstrap/Form'
-      {subLink.label}
+  const getNavListLink = (link: NavLink, index: number) => (
+    <NavDropdown.Item
+      className={link.className ? link.className.concat(' ', '') : ''}
```
I don't think that we need any of the concatenation stuff here. It should be sufficient to do `className={link.className}`

Same for all of the other ones, too.

ocBruno commented on scss/_variables.scss:
```diff
@@ -6,5 +6,5 @@ $theme-colors: (
   "info": #177db8,
   "light": #fdfffc,
-  "dark": #011627
+  "dark": #082338
```
Hey @jackcmeyer, if this is all that's needed to close this request give me a heads up and I can leave the old color until further notice so I can proceed with another conflicting pull request that makes the navbar more mobile friendly.

jackcmeyer replied:
Sounds good. Let’s revert back to the old color.
